### PR TITLE
use custom parser for gts/gjs

### PR DESCRIPTION
**Pull request by patricklx**
bonus:
 * enables type aware lints
 * prettier eslint plugin (with template tag prettier plugin) will just work for gts/gjs
 * can detect unused block params in templates
 * can detect undef vars in PathExpression
 * can add eslint directive comments in mustache or html disadvantage:
* prettier will not work without template tag prettier plugin for gts/gjs files

disavdvantage:
needs specific configuration to enable the parser, as setting it globally will overwrite it.
like https://github.com/vuejs/vue-eslint-parser does.

Breaking change:
* semi rules for template part. by coincidence same as prettier.
* no-undef rule will take effect for template vars (includes js scope)
* no-unsed rule will take effect for template block params
* new config for gjs/gts, notified if not setup correctly 

e.g.
```js
// .eslintrc.js
module.exports = {
  plugins: ['ember'],
  extends: [
    'eslint:recommended',
    'plugin:ember/recommended', // or other configuration
  ],
  overrides: [
    {
      files: ['**/*.gts', '**/*.gjs'],
      parser: 'eslint-plugin-ember/gjs-gts-parser'
    },
    }
  ],
  rules: {
    // override / enable optional rules
    'ember/no-replace-test-comments': 'error'
  }
};
```

fixes https://github.com/ember-cli/eslint-plugin-ember/issues/1770
fixes https://github.com/ember-cli/eslint-plugin-ember/issues/1894
fixes https://github.com/ember-cli/eslint-plugin-ember/issues/1747
fixes https://github.com/ember-cli/eslint-plugin-ember/issues/1683
fixes https://github.com/ember-cli/eslint-plugin-ember/issues/1659
also fixes https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/81
fixes https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/84
closes https://github.com/ember-cli/eslint-plugin-ember/pull/1667
closes https://github.com/ember-cli/eslint-plugin-ember/issues/1957

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,6 +42,29 @@ module.exports = {
...
+  extends: [
+    'eslint:recommended',
+    'plugin:@typescript-eslint/base', // or typescript-eslint/recommended, must come before plugin:ember/recommended
+    'plugin:ember/recommended', // or other configuration
+  ],
+  rules: {
+    // override / enable optional rules
+    'ember/no-replace-test-comments': 'error'
+  }
+};
```
Comment (NullVoxPopuli) [inline README.md:65]: I'd prefer we specify: 
```
Overrides: [
{
  files: ["**/*.{gjs,gts}"],
  parser: 'ember',
}
]
```

Comment (patricklx) [reply README.md:65]: would this also be okay?
```js
{
      files: ['**/*.gts'],
      parser: 'eslint-plugin-ember/gts-parser',
    },
    {
      files: ['**/*.gjs'],
      parser: 'eslint-plugin-ember/gjs-parser',
    },
```

**Comment by patricklx** [comment]
@NullVoxPopuli @bmish i just pushed a non-breaking version.
this makes the parser strategy opt-in

File: README.md (hunk 1 above)
Comment (NullVoxPopuli) [reply README.md:65]: Yeah

Comment (patricklx) [reply README.md:73]: @NullVoxPopuli 
with the non-breaking approach I also need to overwrite the preprocessor:
so better:
```js
overrides: [
    {
      files: ['**/*.gts'],
      extends: [
        'eslint:recommended-gjs',
      ]
    },
    {
      files: ['**/*.gjs'],
      extends: [
        'eslint:recommended-gjs',
      ]
    }
  ]
  ```

**Comment by patricklx** [comment]
maybe we can then add a deprecation warning into the glimmer preprocesor. and later even error out to inform the user about the required changes

**Comment by patricklx** [comment]
@bmish @NullVoxPopuli 
I think there are two options
1. Use non breaking version and keep old code
2. breaking version, and we keep the preprocessor which only throws an error notifying the user how to setup the parser.

File: lib/preprocessors/noop.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  },
+  preprocess: null,
+  postprocess: (messages, fileName) => {
+    const msgs = messages.flat();
+    if (
+      msgs.length === 1 &&
+      msgs[0].ruleId === null &&
+      msgs[0].message.startsWith('Parsing error: ')
+    ) {
+      if (!parsedFiles.has(fileName)) {
```
Comment (patricklx) [inline lib/preprocessors/noop.js:15]: this is how we check if the file went through our parser or not.
I only notify the user if a parsing error happened. We could also always notify the user if its not correctly setup.

File: tests/lib/rules-preprocessor/gjs-gts-parser-test.js [hunk 3]
```diff
@@ -134,22 +156,48 @@ const valid = [
...
+    parser: '@typescript-eslint/parser',
+    filename: 'my-component.gjs',
+    code: `import Component from '@glimmer/component';
+    export default class MyComponent extends Component {
+      <template>Hello!</template>
+    }`,
+    errors: [
+      {
+        message:
+          'Parsing error: Unexpected token. A constructor, method, accessor, or property was expected.\n' +
```
Comment (patricklx) [inline tests/lib/rules-preprocessor/gjs-gts-parser-test.js:182]: here is the error when not setting it up correctly

File: tests/lib/rules-preprocessor/gjs-gts-parser-test.js [hunk 4]
```diff
@@ -134,22 +156,48 @@ const valid = [
...
+          'To lint Gjs/Gts files please follow the setup guide at https://github.com/ember-cli/eslint-plugin-ember#gtsgjs-eslint-setup',
```
Comment (patricklx) [inline tests/lib/rules-preprocessor/gjs-gts-parser-test.js:183]: here is the error when not setting it up correctly

**Comment by patricklx** [comment]
@bmish @NullVoxPopuli 
I found a way to notify the users if their eslint config needs to be updated for gts/gjs and link them to the README.md.
Or should I include the setup in the error message?

**Comment by NullVoxPopuli** [comment]
In your PR description, your eslint config uses the root namespace for configuring ember stuff -- I think we need to move the community more towards an [overrides-only approach](https://github.com/embroider-build/addon-blueprint/issues/71), which I'd been trying to get folks to care about for  some time.

For example:
```js
// .eslintrc.js
module.exports = {
  overrides: [
    {
      files: ['**/*.gts'],
      parser: 'eslint-plugin-ember/gts-parser',
      plugins: ['ember'],
      extends: [
        'eslint:recommended',
        'plugin:ember/recommended', // or other configuration
      ],
    },
    {
      files: ['**/*.gjs'],
      parser: 'eslint-plugin-ember/gjs-parser',
      plugins: ['ember'],
      extends: [
        'eslint:recommended',
        'plugin:ember/recommended', // or other configuration
      ],
    },
    {
      files: ['tests/**/*.{gjs,gts}'],
      rules: {
        // override / enable optional rules
        'ember/no-replace-test-comments': 'error'
      }
    }
  ],
};
```

Reason being that it's way easier to manage long term, as it flattens the amount of cascading you force yourself in to, and you can more easily opt out of plugins, because they never existed for a particular folder in the first place.

This aligns nicely with the direction ESLint9 is going in as the current config format will be dropped. 🙃 


_this is all tangential to the PR tho_ lol

File: lib/preprocessors/noop.js (hunk 2 above)
Comment (NullVoxPopuli) [reply lib/preprocessors/noop.js:15]: I think this is great! is there a way to check if a gjs / gts file was passed to this though? right now I have a concern about non-gjs and non-gts files hitting this message, would be over-aggressive

File: tests/lib/rules-preprocessor/gjs-gts-parser-test.js [hunk 5]
```diff
@@ -472,6 +564,40 @@ describe('lint errors on the exact line as the <template> tag', () => {
   });
 });
 
+describe('supports eslint directives inside templates', () => {
```
Comment (NullVoxPopuli) [inline tests/lib/rules-preprocessor/gjs-gts-parser-test.js:567]: oooooo

File: tests/lib/rules-preprocessor/gjs-gts-parser-test.js [hunk 6]
```diff
@@ -472,6 +564,40 @@ describe('lint errors on the exact line as the <template> tag', () => {
...
+
+    const resultErrors = results.flatMap((result) => result.messages);
+    expect(resultErrors).toHaveLength(0);
+  });
+  it('works with html comment', async () => {
+    const eslint = initESLint();
+    const code = `
+    <template>
+      <div>
+        <!--eslint-disable-next-line-->
```
Comment (NullVoxPopuli) [inline tests/lib/rules-preprocessor/gjs-gts-parser-test.js:589]: wouldn't this get shipped to users tho? 😅 
(non-blocker, just a goofiness)

**Review by NullVoxPopuli — APPROVED** [review]
Nice work -- thanks for iterating on this, and greatest apologies we had to revert the first time around <3

File: lib/preprocessors/noop.js (hunk 2 above)
Comment (patricklx) [reply lib/preprocessors/noop.js:15]: Well, the preprocessor is only setup for those files

**Comment by patricklx** [comment]
> In your PR description, your eslint config uses the root namespace for configuring ember stuff -- I think we need to move the community more towards an [overrides-only approach](https://github.com/embroider-build/addon-blueprint/issues/71), which I'd been trying to get folks to care about for some time.
> 
> For example:
> 
> ```js
> // .eslintrc.js
> module.exports = {
>   overrides: [
>     {
>       files: ['**/*.gts'],
>       parser: 'eslint-plugin-ember/gts-parser',
>       plugins: ['ember'],
>       extends: [
>         'eslint:recommended',
>         'plugin:ember/recommended', // or other configuration
>       ],
>     },
>     {
>       files: ['**/*.gjs'],
>       parser: 'eslint-plugin-ember/gjs-parser',
>       plugins: ['ember'],
>       extends: [
>         'eslint:recommended',
>         'plugin:ember/recommended', // or other configuration
>       ],
>     },
>     {
>       files: ['tests/**/*.{gjs,gts}'],
>       rules: {
>         // override / enable optional rules
>         'ember/no-replace-test-comments': 'error'
>       }
>     }
>   ],
> };
> ```
> 
> Reason being that it's way easier to manage long term, as it flattens the amount of cascading you force yourself in to, and you can more easily opt out of plugins, because they never existed for a particular folder in the first place.
> 
> This aligns nicely with the direction ESLint9 is going in as the current config format will be dropped. 🙃
> 
> _this is all tangential to the PR tho_ lol

That would also be good, by using extends in overrides it would not also not be strictly necessary to specify the parser. As that is set in the recommended config.
So only one config would be needed:
```js
{
      files: ['**/*.{gjs,gts}'],
      parser: 'eslint-plugin-ember/gjs-parser',
        plugins: ['ember'],
       extends: [
         'eslint:recommended',
         'plugin:ember/recommended', // or other 
}
```

**Comment by patricklx** [comment]
follow-up prs i would like to include in the next major:
* #1944 
* #1943 
* #1939

**Comment by patricklx** [comment]
When content-tag supports lint mode transform, the relevant code parts can be updated in a later pr. That should be a small change.
It looks like content tag will have the same ouput and meta as ember-template-tag?

**Comment by patricklx** [comment]
I can make the update to this repo once content-tag has the changes.
@bmish any comments?

File: README.md (hunk 1 above)
Comment (bmish) [reply README.md:69]: This seems fine, but just wondering, is there a need or benefit to separating the parsers, or can they be specified as one? 

```
overrides: [
  {
    files: ["**/*.{gjs,gts}"],
    parser: 'eslint-plugin-ember/gjs-gts-parser',
  }
]
```

File: lib/preprocessors/noop.js
Comment (bmish) [inline lib/preprocessors/noop.js:1]: Can you add a high-level comment at the top of this file explaining what this preprocessor is for.

File: README.md [hunk 7]
```diff
@@ -42,6 +42,33 @@ module.exports = {
 };
 ```
 
+## Gts/Gjs eslint setup
```
Comment (bmish) [inline README.md:45]: Simpler title?

```suggestion
## gts/gjs
```

Comment (bmish) [inline README.md:45]: Can you add a few high-level sentences explaining why someone should bother with this Gts/Gjs setup? For someone who isn't familiar with Gts/Gjs, what is it? How can they learn more about it? Why should anyone care?

File: README.md [hunk 8]
```diff
@@ -42,6 +42,33 @@ module.exports = {
...
+    },
+    {
+      files: ['**/*.gjs'],
+      parser: 'eslint-plugin-ember/gjs-parser',
+    }
+  ],
+  rules: {
+    // override / enable optional rules
+    'ember/no-replace-test-comments': 'error'
+  }
```
Comment (bmish) [inline README.md:68]:
```suggestion
```

File: README.md [hunk 9]
```diff
@@ -42,6 +42,33 @@ module.exports = {
...
+## Gts/Gjs eslint setup
+
+```js
+// .eslintrc.js
+module.exports = {
+  plugins: ['ember'],
+  extends: [
+    'eslint:recommended',
+    'plugin:ember/recommended', // or other configuration
+  ],
```
Comment (bmish) [inline README.md:54]: Let's only include the new `overrides` instead of repeating the parts of the config already present above.

Comment (bmish) [inline README.md:54]: Let's only include the new `overrides` instead of repeating the parts of the config already present above.
```suggestion
```

File: lib/preprocessors/noop.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        msgs[0].message += '\n';
+        msgs[0].message +=
+          'To lint Gjs/Gts files please follow the setup guide at https://github.com/ember-cli/eslint-plugin-ember#gtsgjs-eslint-setup';
```
Comment (bmish) [inline lib/preprocessors/noop.js:18]: When is this message going to show up? I do not want to nag all users about a feature they may not care about. Shouldn't it only show up when they have Gjs/Gts files without specifying the override for them?

File: README.md (hunk 1 above)
Comment (patricklx) [reply README.md:69]: we could.
we would need to detect if the extension is gts or gjs and choose by that, which internal parser should be chosen

File: lib/config/recommended.js
Comment (bmish) [inline lib/config/recommended.js:1]: You mentioned breaking changes in various comments. Can you summarize the status in this thread? I want to confirm this PR is not a breaking change. What is the impact of this PR to existing users today? There should be no change unless opted-in using the override, right? What future breaking changes do you want to make?

File: README.md (hunk 1 above)
Comment (bmish) [reply README.md:69]: Would that be simple? Is there any reason why we shouldn't just auto-detect the extension for the user? If the user doesn't care about distinguishing between these file types, then we can avoid exposing that complexity to the user, and just handle it internally, right?

File: lib/preprocessors/noop.js (hunk 10 above)
Comment (patricklx) [reply lib/preprocessors/noop.js:18]: yes, only shows up for gts&gjs files when not correctly setup

File: lib/preprocessors/noop.js
Comment (patricklx) [reply lib/preprocessors/noop.js:1]: done

File: lib/config/recommended.js
Comment (patricklx) [reply lib/config/recommended.js:1]: this is a breaking change for those having gjs/gts files.
they need to change their config, and if they do not, they will be notified via the preprocessor.postprocess

File: README.md (hunk 7 above)
Comment (NullVoxPopuli) [reply README.md]: do you mean like, links to the RFC, a brief blurb about Polaris, etc?

File: lib/config/recommended.js
Comment (patricklx) [reply lib/config/recommended.js:1]: @bmish what also changes is
* semi rules for template part. by coincidence same as prettier.
* no-undef rule will take effect for template vars
* no-unsed rule will take effect for template block params

**Comment by patricklx** [comment]
@bmish any news about this. I updated thr description to describe the breaking changes

File: README.md (hunk 7 above)
Comment (bmish) [reply README.md]: Yes, a one-sentence overview and links to docs/guides helps.

File: lib/config/recommended.js
Comment (bmish) [reply lib/config/recommended.js:1]: So it sounds like there are breaking changes for mainstream users, meaning this PR would require a major release. We have an issue tracking the next major release https://github.com/ember-cli/eslint-plugin-ember/issues/1535 and it will take some work and time to get that out (I don't have much time these days).

Just to confirm, can you describe any non-breaking-change alternative approaches to releasing these changes, and let us know which approach you recommend? Brand new functionality as a breaking change can be risky since the community hasn't had an opt-in chance to test it yet.

Comment (patricklx) [reply lib/config/recommended.js:1]: Well, we could keep the old preprocessor and just have the override for the new code set the new preprocessor:
```
overrides: [
    {
      files: ['**/*.gts', '**/*.gjs'],
      parser: 'eslint-plugin-ember/gjs-gts-parser',
      preprocessor: 'ember/noop'
    },
```
But i suggest to just use the parser without old code, as the goal here is to also remove ember-template-imports. 
And using another tool in the preprocessor which generates the new syntax will only cause more problems.
Combined with the fact that gts is not yet fully official , considering that it's not yet in the [release state](https://github.com/emberjs/rfcs/pull/871)
I think this is risk we can take.

Comment (bmish) [reply lib/config/recommended.js:1]: Are you suggesting this shouldn't be a breaking change? Can you be clear about who this is a breaking change for?

Comment (patricklx) [reply lib/config/recommended.js:1]: This is a breaking change for everyone using gts/gjs files.
I'm just saying that they should be aware of some risks, as gts not fully released yet.

**Comment by patricklx** [comment]
@bmish any news for this?
Ember template imports 4 has been released.

**Comment by NullVoxPopuli** [comment]
eti v4 is more a signal that content-tag should be good to go in the linting prs.

**Comment by patricklx** [comment]
content tag still doesn't emit a format that's usable for linting

**Comment by NullVoxPopuli** [comment]
It has a second api specifically for linting. Check my glint pr where the ranges are used

**Comment by patricklx** [comment]
I will have a look, but here ranges are not enough.
Need to have a transform  which doesn't change offsets

**Comment by patricklx** [comment]
mmm, so glint uses that to do it's own transform?
it's also possible to know if the template is inside a class expression or normal expression?
If so, then i could also do the custom transform here...

**Review by NullVoxPopuli — APPROVED** [review]
Looks great!

**Comment by NullVoxPopuli** [comment]
@patricklx this one needs rebase real quick

File: lib/parsers/gjs-gts-parser.js [hunk 11]
```diff
@@ -0,0 +1,498 @@
+const ContentTag = require('content-tag');
+const glimmer = require('@glimmer/syntax');
+const DocumentLines = require('../utils/document');
+const { visitorKeys: glimmerVisitorKeys } = require('@glimmer/syntax');
```
Comment (patricklx) [inline lib/parsers/gjs-gts-parser.js:4]: yes, no hacky require anymore 🚀

Comment (NullVoxPopuli) [reply lib/parsers/gjs-gts-parser.js:4]: What version is in use here?
Current release of glimmer/syntax is ESM only 😅

File: package.json [hunk 12]
```diff
@@ -65,26 +68,28 @@
     ]
   },
   "dependencies": {
+    "@babel/eslint-parser": "^7.15.8",
     "@ember-data/rfc395-data": "^0.0.4",
-    "@glimmer/syntax": "^0.84.2",
+    "@glimmer/syntax": "^0.85.5",
```
Comment (NullVoxPopuli) [inline package.json:73]: Hm, this is latest. I guess require can work with an esm module in some situation.

Comment (patricklx) [reply package.json:73]: mmm, actually didnt test the code...

Comment (NullVoxPopuli) [reply package.json:73]: Is there an async location to import this?

Otherwise i wonder if a whole eslint plugin can be esm

Comment (patricklx) [reply package.json]: I think esm eslint plugin can only be used with the new config files.

**Comment by patricklx** [comment]
@NullVoxPopuli this now uses latest glimmer/syntax

**Review by NullVoxPopuli — APPROVED** [review]
